core/src/unprocessed_packet_batches.rs

```diff
@@ -134,6 +181,188 @@ impl UnprocessedPacketBatches {
+        let mut fee_buckets = BTreeMap::<u64, Vec<PacketLocator>>::new();
+        for locator in locators {
+            let fee_per_cu = if let Some(fee_per_cu) = self.compute_fee_per_cu(locator, &bank) {
+                fee_per_cu
+            } else {
+                // if unable to compute fee-per-cu for the packet, put it to the `0` bucket
+                0u64
+            };
+
+            let bucket = fee_buckets
+                .entry(fee_per_cu)
+                .or_insert(Vec::<PacketLocator>::new());
+            bucket.push(locator.clone());
+        }
+        fee_buckets
+            .iter()
+            .rev()
+            .flat_map(|(_key, bucket)| bucket.iter().cloned())
+            .collect()
+    }
+
+    /// Computes `(addition_fee + base_fee / requested_cu)` for packet is referenced by `PacketLocator`
+    fn compute_fee_per_cu(&self, locator: &PacketLocator, bank: &Option<Arc<Bank>>) -> Option<u64> {
+        if let Some(bank) = bank {
+            let deserialized_packet_batch = self.get(locator.batch_index)?;
+            let deserialized_packet = deserialized_packet_batch
+                .unprocessed_packets
+                .get(&locator.packet_index)?;
+            let sanitized_message = Self::sanitize_message(
+                &deserialized_packet.versioned_transaction.message,
+                bank.as_ref(),
+            )?;
+            let total_fee = bank.get_fee_for_message(&sanitized_message)?;
+
+            // TODO refactor `bank.get_fee_for_message()` to return both fee and CUs to avoid
+            // calling ComputeBudget twice.
+            let mut compute_budget = ComputeBudget::default();
+            let _ = compute_budget
+                .process_message(&sanitized_message, false)
+                .ok()?;
+
+            Some(total_fee / compute_budget.max_units)
+        } else {
+            None
+        }
+    }
+
+    fn sanitize_message(
+        versioned_message: &VersionedMessage,
+        address_loader: impl AddressLoader,
+    ) -> Option<SanitizedMessage> {
+        versioned_message.sanitize().ok()?;
+
+        match versioned_message {
+            VersionedMessage::Legacy(message) => Some(SanitizedMessage::Legacy(message.clone())),
+            VersionedMessage::V0(message) => {
+                let loaded_addresses = address_loader
+                    .load_addresses(&message.address_table_lookups)
+                    .ok()?;
+                Some(SanitizedMessage::V0(v0::LoadedMessage::new(
+                    message.clone(),
+                    loaded_addresses,
+                )))
+            }
+        }
+    }
+
+    /// This function is called to put new packet_batch into buffer when it is at Max capacity.
+    /// It tries to drop lower prioritized transactions in order to find an empty batch to swap
+    /// with new packet_batch.
+    /// Returns the dropped packet_batch, dropped batch count and dropped packets count.
+    fn swap_packet_by_priority(
```

Comment:
nit: maybe rename to `drop_packets_by_priority`

Reply:
maybe `replace` instead of `drop`?

Reply:
that works for me!

Reply:
👍🏼